### PR TITLE
Create openstack chart

### PR DESCRIPTION
**Pull request by petesfrench**
## Done

- Creates pie-chart representing number of commit by company to openstack

## QA

- Check out this feature branch
- Run the site using the command `./run serve` or `dotrun`
- View the site locally in your web browser at: http://0.0.0.0:8001/openstack/what-is-openstack
- Run through the following [QA steps](https://canonical-web-and-design.github.io/practices/workflow/qa-steps.html)
- Compare against [copy-doc](https://docs.google.com/document/d/15xrgIq_KUwHaCyFoI7-yS_JmBbsoyXJ4GpRMFAa7_is/edit?ts=60a3e5c8#)

## Issue / Card

Fixes https://github.com/canonical-web-and-design/ubuntu.com/issues/9782

### CODE REVIEW

**Comment by anthonydillon** [comment]
I'm off for a week I'm afraid. Mind if this could be assigned to someone else for review.

**Comment by spencerbygraves** [comment]
@anthonydillon @anasereijo have we done anything similar to this pie chart before? We'll need to look at the colours and hover state I think.

File: templates/openstack/what-is-openstack.html [hunk 1]
```diff
@@ -64,6 +64,140 @@ <h2>Companies involved in OpenStack</h2>
...
+
+        var tots = d3.sum(data, function(d) {
+                return d.metric;
+            });
+
+        data.forEach(function(d) {
+            d.percentage = Math.round((100 * (d.metric / tots)) * 10) / 10;
+        });
+
+        var div = d3.select("body").append("div")
```
Comment (steverydz) [inline templates/openstack/what-is-openstack.html:122]: could this variable reflect the name of the part of the chart it is?

File: templates/openstack/what-is-openstack.html [hunk 2]
```diff
@@ -64,6 +64,140 @@ <h2>Companies involved in OpenStack</h2>
...
+                    .style("z-index", "10")
+                    .style("position", "absolute")
+                    .style("background-color", "white")
+                    .style("border", "solid")
+                    .style("border-width", "1px")
+                    .style("border-radius", "3px")
+                    .style("padding", "3px")
+                    .style("pointer-events", "none");
+
+        var g = svg.append("g")
```
Comment (steverydz) [inline templates/openstack/what-is-openstack.html:134]: could this variable reflect the name of the part of the chart it is?

**Review by steverydz — CHANGES_REQUESTED** [review]
The code looks good. I've left a couple of comments asking for some more descriptive names on variables.

I think this code should be in it's own file and ran through the build pipeline. Doing that would also allow you to import only the parts of d3 that you need.

**Comment by steverydz** [comment]
After thinking about it I have a couple more comments:
1. You could add `tabindex="0"` to each section of the pie chart. This will allow keyboard users to tab to each section and read the title
2. Even without needing a screen reader it's not very clear which section applies to which company. Perhaps a key would be useful here?

**Comment by petesfrench** [comment]
@steverydz Thank you for the feedback, I think that's a really good idea to use tabIndex="0". Not sure what we are doing regarding using a key, still waiting on feedback from design.

**Comment by anasereijo** [comment]
Hey, @petesfrench this is working great!

I have some visual suggestions:

- We built [this page](https://ubuntu.com/desktop/statistics) (ages ago) - can you please use the same colours? I know the graphs are just using orange, aubergine and grey but maybe we can try using whatever we have for the world map? Not sure if I remember well, but I think we picked 2 colours - a dark and a lighter one - and the ones in between are created automatically? Anyway, I'm happy to discuss this with you or pick 9 colours.

- We should use the [Vanilla tooltip](https://vanillaframework.io/docs/patterns/tooltips), if possible.

- can you please close my comments with links to illustrations/logos if you already copied them?

Let me know if it's easier to jump on a call and go through this!
Thanks

**Comment by anasereijo** [comment]
As discussed with @petesfrench I am turning the table and pie chart into a bar chart. This work is being tracked [here](https://github.com/canonical-web-and-design/web-squad/issues/4166)

**Comment by steverydz** [comment]
The pie chart is still covered on mobile. I think setting an explicit width and height on it will solve that.

**Comment by anasereijo** [comment]
@petesfrench is this ready for a design review?

**Comment by anthonydillon** [comment]
Red hat is off the chart!

File: templates/openstack/what-is-openstack.html [hunk 3]
```diff
@@ -317,9 +362,20 @@ <h2>Companies involved in OpenStack</h2>
 <section class="p-strip--light">
   <div class="row">
     <div class="col-8">
-      <p>Canonical is the founding member of the foundation, official maintainer and publisher of OpenStack packages on Ubuntu, leader of the <a class="p-link--external" href="https://docs.openstack.org/charm-guide/latest/">OpenStack Charms project</a> and one the biggest contributor to OpenStack all the time.</p>
+      <p>Canonical is the founding member of the foundation, official maintainer and publisher of OpenStack packages on
```
Comment (steverydz) [inline templates/openstack/what-is-openstack.html:365]: I've made a couple of suggestions in the copy

File: static/js/src/openstack-chart.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  var chartHeight = viewHeight - margin.top - margin.bottom;
+  var fontSize = viewWidth / 52;
+
+  var svg = d3
+    .select("#openstack-bar-chart")
+    .append("svg")
+    .attr("class", "svg")
+    .attr("width", viewWidth)
+    .attr("height", viewHeight);
+  // .attr("viewbox", "0 0" + viewWidth + viewHeight);
```
Comment (steverydz) [inline static/js/src/openstack-chart.js:35]: Remove commented out code

File: static/js/src/openstack-chart.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+function drawChart(data) {
+  var totals = d3.sum(data, function (d) {
+    return d.metric;
+  });
+  data.forEach(function (d) {
+    d.percentage = Math.round(100 * (d.metric / totals) * 10) / 10;
+  });
+  data.reverse();
+
+  var aspectRatio = 0.55;
```
Comment (steverydz) [inline static/js/src/openstack-chart.js:18]: These `var` can all be `const`

File: static/js/src/openstack-chart.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
+fetch("/static/js/data/openstack-commit-stats.json")
+  .then((response) => response.json())
+  .then((data) => {
+    data = data.stats.slice(0, 10);
+    drawChart(data);
+    drawTable(data);
+  });
+
+function drawChart(data) {
+  var totals = d3.sum(data, function (d) {
```
Comment (steverydz) [inline static/js/src/openstack-chart.js:10]: This `var` can be `const`

Comment (steverydz) [inline static/js/src/openstack-chart.js:10]: You can use a arrow function here

File: static/js/src/openstack-chart.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    return d.metric;
+  });
+  data.forEach(function (d) {
```
Comment (steverydz) [inline static/js/src/openstack-chart.js:13]: You can use a arrow function here

File: static/js/src/openstack-chart.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+function drawTable(data) {
+  const tableHeadings = `
+    <thead>
+      <tr>
+        <th>Company</th>
+        <th>Commits</th>
+      </tr>
+    </thead>`;
+  let tableContent = ``;
+  for (let i = 0; i < data.length; i++) {
```
Comment (steverydz) [inline static/js/src/openstack-chart.js:93]: you can use `forEach` here instead of a `for` loop

**Review by steverydz — CHANGES_REQUESTED** [review]

File: templates/openstack/what-is-openstack.html [hunk 9]
```diff
@@ -505,7 +569,14 @@ <h2>Who uses OpenStack?</h2>
   <div class="row">
     <div class="col-8">
       <h2>How big is the OpenStack project?</h2>
-      <p>According to the <a class="p-link--external" href="https://451research.com/search?products=OpenStack">Research Market Monitor (Open Source Software, OpenStack)</a> from September 2019, OpenStack’s combined market size worldwide is $7.7B. It also continues to be one of the top three open source projects in the world in terms of active contributions, along with the Linux kernel and Chromium. In 2020 OpenStack reached its 10-year anniversary which proves its maturity as a software project. Governed by the <a class="p-link--external" href="https://openinfra.dev/">Open Infrastructure Foundation</a>, more than 34,000 individual contributors and 550 companies participate in OpenStack development. It is also the most popular open source private cloud platform and its adoption continues to grow.</p>
+      <p>According to the <a class="p-link--external" href="https://451research.com/search?products=OpenStack">Research
```
[the remark this answers is not in the document]
Comment (petesfrench) [reply templates/openstack/what-is-openstack.html:572]: Not sure what the suggested change is here

[the remark this answers is not in the document]
Comment (steverydz) [reply templates/openstack/what-is-openstack.html:572]: The link text of the link doesn't match the copy doc. In the copy doc it starts with "451"

**Comment by petesfrench** [comment]
Thanks @steverydz, I made those changes so is ready for another review :+1:

**Review by steverydz — APPROVED** [review]
LGTM, but every instance of the image template seems to have been reformatted to one line? Could you change that back to multiline so it's easier to work with? Other than that, all good!

**Comment by tytus-kurek** [comment]
@anasereijo this is not compelling anymore. The pie chart from stackanalytics.io includes the "others" field which makes Red Hat's portion smaller and the bar-based approach makes the difference even more significant.

I'd suggest we do the following:
- We mask other contributors with something like "Contributor I", "Contributor II", etc.
- We only mark the Canonical bar aubergine and mark the other bars grey or something like that
- We add the source under the diagram (https://www.stackalytics.io/?release=all&metric=commits) so that when people click on it they see "others" as the biggest contributor and they think it's this 35.7% one.

Those should do the trick.

**Review by tytus-kurek — CHANGES_REQUESTED** [review]
Comments above.

**Comment by anasereijo** [comment]
@tytus-kurek sounds good to me, thanks for your review. 

@petesfrench let me know If you need my help on this

**Comment by petesfrench** [comment]
@anasereijo I made those changes with some design assumptions, can you take a look and let me know what you think

**Comment by anasereijo** [comment]
Thanks, @petesfrench for making the changes!
please use the `#666666` grey and make Canonical `font-weight: 400;`

I asked @davegoddard42 what he thinks about just using `Source` as a link, let's see what he says :)

**Comment by anasereijo** [comment]
> Thanks, @petesfrench for making the changes!
> please use the `#666666` grey and make Canonical `font-weight: 400;`
> 
> I asked @davegoddard42 what he thinks about just using just the word `Source` as a link, let's see what he says :)

@peterfrench, please change the link to: `Source: stackalytics` as per @davegoddard42's suggestion. Thanks!

**Comment by SirSamTumless** [comment]
@tytus-kurek Please re-review this

**Comment by tytus-kurek** [comment]
@SirSamTumless LGTM. I have a problem assigning `PdM needed` label.

**Comment by SirSamTumless** [comment]
@petesfrench This is all good to go 🙌

**Pull request merged**